internal/api/impl/v1/dashboard/validator/validator.go

```diff
@@ -0,0 +1,220 @@
+/*
+ * Validate a dashboard.
+ * The dashboard's panels are matched against the known list of CUE definitions (schemas).
+ * If no schema matches, the validation fails.
+ */
+func (v *validator) Process(dashboard *v1.Dashboard) error {
+	logrus.Tracef("Dashboard to validate: %+v", dashboard)
+
+	// init the validator if it wasn't already
+	if s, _ := v.baseDef.String(); len(s) == 0 {
+		v.initSchemas()
+	}
+
+	var res error
+
+	// go through the panels list
+	// the processing stops as soon as it detects an invalid panel TODO: can probably be improved
+	for k, panel := range dashboard.Spec.Panels {
+		logrus.Tracef("Panel to validate: %s", string(panel))
+
+		// compile the JSON panel into a CUE Value
+		value := v.context.CompileBytes(panel)
+
+		// retrieve panel's kind
+		kind, err := value.LookupPath(cue.ParsePath("kind")).String()
+		if err != nil {
+			logrus.Error("This panel doesn't contain the required \"kind\" field")
+			res = err
+			break
+		}
+
+		// retrieve the corresponding schema
+		schema, ok := v.schemas.Load(kind)
+		if !ok {
+			logrus.Errorf("%s is not valid panel: unknown kind %s", k, kind)
+			res = errors.New("Unknown panel kind")
+			break
+		}
+		logrus.Tracef("Matching panel %s against schema: %+v", k, schema)
+
+		// do the validation
+		unified := value.Unify(schema.(cue.Value))
+		opts := []cue.Option{
+			cue.Concrete(true),
+			cue.Attributes(true),
+			cue.Definitions(true),
+			cue.Hidden(true),
+		}
+		err = unified.Validate(opts...)
+		if err != nil {
+			logrus.Errorf("%s is not a valid %s panel", k, kind)
+			res = err
+			break
+		}
+	}
+
+	if res == nil {
+		logrus.Debug("This dashboard is OK, all its panels are valid")
+	} else {
+		logrus.WithError(res).Error("This dashboard is KO, at least 1 panel is invalid")
+	}
+
+	return res
+}
+
+/*
+ * Load the known list of schemas into the validator
+ */
+func (v *validator) LoadSchemas(path string) {
+	schemas, err := loadSchemas(v.context, v.baseDef, fmt.Sprintf("%s/%s", path, pluginsFolderName))
+	if err != nil {
+		logrus.WithError(err).Error("Not able to retrieve the list of schema files")
+		return
+	}
+
+	v.schemas = schemas
+	logrus.Info("Schemas list (re)loaded")
+}
+
+/*
+ * Load & return the known list of schemas
+ */
+func loadSchemas(context *cue.Context, baseDef cue.Value, path string) (*sync.Map, error) {
+	var schemas sync.Map
+
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		return &schemas, err
+	}
+
+	// process each schema plugin to convert it into a CUE Value
+	// for each schema we check that it meets the default specs we expect for any panel, otherwise we dont include it
+	for _, file := range files {
+		schemaPath := fmt.Sprintf("%s/%s", path, file.Name())
+
+		// load the cue files into build.Instances slice
+		buildInstances := load.Instances([]string{}, &load.Config{Dir: schemaPath})
+		// we strongly assume that only 1 buildInstance should be returned (corresponding to the main definition like #panel), otherwise we skip it
```

Comment:
do we assume here that only one main definition should be present per cue package ?
Is it a best practice in Cuelang to proceed like that ?

Reply:
it depends on what you'd want to validate, let's say you wanted to validate Perses config:

```
database:
  etcd:
    protocol: "http"
    request_timeout: 10
    connections:
      - host: 0.0.0.0
        port: 2379

schemas:
  path: "cue/schemas"
```

In this case you have 2 "top" objects schemas and database, so it would make sense to have a cue schema with two definitions #database and #schemas at its root. In our case since we are validating panels one by one, I think it makes sense to enforce that their should be a top #panel definition. Anyway if ever later we find out that it's too restrictive, that's something we could change.